bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -383,7 +383,9 @@ PrePrepareMsg *ReplicaImp::buildPrePrepareMsgBatchByOverallSize(uint32_t require
 }
 
 PrePrepareMsg *ReplicaImp::buildPrePrepareMsgBatchByRequestsNum(uint32_t requiredRequestsNum) {
-  if (requestsQueueOfPrimary.size() < requiredRequestsNum) {
+  ConcordAssertGT(requiredRequestsNum, 0);
+  // DD: To make sure that time service does not affect sending messages
+  if (requestsQueueOfPrimary.size() < requiredRequestsNum - static_cast<uint32_t>(config_.timeServiceEnabled)) {
```

Comment:
Since the Time Service request is created only when `PrePrepare` is built, it is not inserted to `requestsQueueOfPrimary`. It is a problem for cases when there are, let us say, 9 requests in the queue, and `requiredRequestsNum` is 10: according to the original logic, the `PrePrepare` won't be built but if TS is on, we know that we have an additional request. So we decrease `requiredRequestsNum` by one if TS is on.

Reply:
Changed the condition to be more explicit.